### PR TITLE
Add runtime profiler info for AOTDispatcher prologue

### LINKED ISSUE #155721: Add runtime profiler info for AOTDispatcher prologue
If you look at the profiler output from running a tiny compiled model, you can see that there is some cpu overhead that we need to churn through before start executing the compiled artifact itself:

Image

The overhead you can see here includes:

(1) TorchDynamo cache lookup overhead

(2) dynamo's pre-graph bytecode (the small green box)

(3) a bunch of un-named overhead afterwards. This is very likely coming from AOTDispatcher's runtime wrapper, so it would be nice to add a profiler event for it, and confirm that the event covers most of this gap.

**code pointers**

(1) we need to insert an event in the profiler in a low-overhead way. @anijain2305 has some examples of doing this [here](https://github.com/pytorch/pytorch/pull/154769/files#diff-b99ae8eb082046d6353fe4ee536a0d846e1007c7d3b21aefae468ded18fbf13eR4677). This example has some extra convoluted code in it because that particular usage involves inserting the profiler call in bytecode, but I think we can just use that context manager directly. We can probably also gate it by a new `torch._functorch.config.runtime_overhead_profiling` flag if we're worried about overhead.

(1) AOTAutograd's main runtime entrypoint is in this function: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L320. So one source of overhead that we can wrap in a context manager is everything in this function, up until the call to the compiled artifact itself: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L344

(2) that outer `runtime_wrapper` actually calls a few inner runtime wrapper functions. The way this is set up is that there are several `def post_compile` functions defined in this file ([example](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1944)). The one I just linked is the "autograd runtime wrapper" that is invoked whenever we do training. We can probably start by adding an event just for this one for now, and see if these two events are enough to cover the common inference/training case, although eventaully we may want to add events for the other post_compile functions as well (although there is some risk that adding 10 of these will add non-zero overhead).

For the autograd post_compile function, we effectively want to wrap everything in the `forward()` method up until  the inner compiled artifact is invoked [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L2034)

You can generate a similar profile by running:
```
import torch
from torch.profiler import profile, ProfilerActivity

mod = torch.nn.Linear(4, 4, device='cuda')
opt_mod = torch.compile(mod)
x = torch.randn(4, 4, device="cuda")
mod(x)
opt_mod(x)
for _ in range(10):
    opt_mod(x)
with profile(activities=[ProfilerActivity.CPU]) as prof:
    opt_mod(x)
prof.export_chrome_trace("trace.json")
```
and running:
```
# run the script
python tmp.py
# upload the trace.json file. this will give you a download link you can run locally,
so you can stick the .json file in your local `chrome://tracing` window
jf upload trace.json
```

cc @chauhang @penguinwu @zou3519

### PR DESCRIPTION
**Pull request by azahed98**
Fixes #155721

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @xmfan

### CODE REVIEW

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 1]
```diff
@@ -354,6 +369,7 @@ def runtime_wrapper(args: list[Any]):
             try:
                 if grad_enabled:
                     torch._C._set_grad_enabled(False)
+                record_runtime_wrapper_prologue_exit(cm)
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:383]: Hmm at first I was going to ask you to use the context manager directly, so we don't need to worry about carefully making sure that the enter/exit line up in all the various branches.

I guess this is actually a pain to do though, since you need different global state to be toggled in the different branches, and you want to capture that inside of the recorded region

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 2]
```diff
@@ -316,8 +316,22 @@ def _create_runtime_wrapper(
             make_output_handler(info, runtime_metadata, trace_joint)
             for info in runtime_metadata.output_info
         )
+    def record_runtime_wrapper_prologue_enter() -> AbstractContextManager[None]:
+        cm: AbstractContextManager[None] = (
+            torch._C._profiler._RecordFunctionFast("AOTDispatcher Runtime Wrapper Prologue")
+            if torch.autograd.profiler._is_profiler_enabled
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: @anijain2305 I saw that you have an additional `config.record_pre_graph_bytecode_in_traces` that you branch on in your case. Did you add that because you're worried about the extra overhead even in this case where profiling is off? https://github.com/pytorch/pytorch/pull/154769/files#diff-0bb1756725c4426408938314b0c9d3988ae5bf49994892d7038ad7746e209e9fR617

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 3]
```diff
@@ -316,8 +317,22 @@ def _create_runtime_wrapper(
             make_output_handler(info, runtime_metadata, trace_joint)
             for info in runtime_metadata.output_info
         )
+    def record_runtime_wrapper_prologue_enter() -> AbstractContextManager[None]:
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: just a thought - if we are worried about overhead we could just have this function return `None` instead of ` nullcontext()` when we are not profiling. `nullcontext()`

**Review by bdhirsh — APPROVED** [review]
nice work!

File: test/dynamo/test_repros.py [hunk 4]
```diff
@@ -5801,6 +5802,40 @@ def f(x, l):
...
+        mod = torch.nn.Linear(4, 4)
+        opt_mod = torch.compile(mod)
+        x = torch.randn(4, 4)
+        
+        # warm up
+        opt_mod(x)
+
+        # Run compiled op with profiling
+        with profile(activities=[ProfilerActivity.CPU]) as prof:
+            opt_mod(x)
```
Comment (bdhirsh) [inline test/dynamo/test_repros.py]: you can test the `trace_joint == False` case by running with no_grad:
````
with torch.no_grad(), profile(...):
    opt_mod(x)
```

Comment (azahed98) [reply test/dynamo/test_repros.py]: Ty! Added, but put the warmup in the no_grad context as well.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 3 above)
Comment (azahed98) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: Changed so now enter returns and exit accepts `Union[AbstractContextManager[None], None]`

**Comment by azahed98** [comment]
Apologies if you were subscribed - I mad a mistake when rebasing and failed to double check that it was done correctly
